src/Compilers/CSharp/Portable/CodeGen/CodeGenerator.cs

```diff
@@ -401,6 +407,58 @@ private void EmitSequencePointStatement(BoundSequencePointWithSpan node)
             }
         }
 
+        private void EmitSaveSequencePoint(BoundSaveSequencePoint statement)
+        {
+            if (!_emitPdbSequencePoints)
+                return;
+
+            ArrayBuilder<RawSequencePoint> sequencePoints = _builder.SeqPointsOpt;
+            if (sequencePoints is null)
+                return;
+
+            for (int i = sequencePoints.Count - 1; i >= 0; i--)
+            {
+                var span = sequencePoints[i].Span;
+                if (span == RawSequencePoint.HiddenSequencePointSpan)
+                    continue;
+
+                // Found the previous non-hidden sequence point.  Save it.
+                _savedSequencePoints ??= PooledDictionary<object, TextSpan>.GetInstance();
+                _savedSequencePoints.Add(statement.Identifier, span);
+                return;
+            }
+        }
```

Comment:
Is it possible that we produced a "save" with no preceding sequence point? If not, let's add an assertion or unreachable exception.

As a follow-up to that question (assuming that we can tighten the logic), is it possible that we "restore" but we didn't have a corresponding "save"?

Reply:
The answer is that yes, it is possible, though the places where it occurs may be associated with misbehaviors (places where we need to insert a no-op before a hidden sequence point so that a proper sequence point isn't "overwritten" by a hidden sequence point).  I'm looking at three issues that may be related to such misbehavior.

Reply:
As for a restore without a save, no, they should always be paired when generated by lowering, but I imagine it might be possible for a save to end up in unreachable (removed) code and thereby not seen by CodeGenerator.  Also, the save may have failed to find a preceding sequence point. I tried to structure the code here to be resilient to such unusual circumstances.

Reply:
Consider adding an assertion then. It may help us catch those misbehaviors.

Reply:
We have six related open issues here that I'm working on separately.  I'd rather not have the compiler crashing until they are fixed.  I don't know that it is reasonable to ever fix them all, as some of these may be situations in which there is no "previous" sequence point by design so there is no need to restore anything (e.g. expression-bodied method with a switch expression as the body).

---
In reply to: [398878280](https://github.com/dotnet/roslyn/pull/42711#discussion_r398878280) [](ancestors = 398878280)